UART Driver

YashrajN opened the pull request:
Put the UART Driver repo in a folder and am making a PR to this repo

YashrajN committed:
Added UART Driver

YashrajN committed:
Added UART Driver

YashrajN committed:
Added the UART Driver

antholuo commented on UART Driver/Core/Src/main.c:
```diff
@@ -0,0 +1,307 @@
+	float roll;
+};
+
+//Should be 37 bytes but because of padding its really 48 bytes
```
I'm OOTL, could you explain the padding situation?

YashrajN replied:
Sure, here's the gist of it: in this case, a 64bit processor can access 8 bytes at a time, so in this example, when reading the foji struct my first byte would be my start, and the next data type in my struct is a double. But the processor only has access to 7 bytes left since the start already took one up. Since a double is 8 bytes it cant fit. This means 7 bytes would be left empty. Then all the doubles would be accessed smoothly without empty bytes because they are all 8 bytes. Then when reading the floats the first 2 floats would be read without empty bytes as they sum up to 8. The last float would be 4 bytes which means 4 empty left. This means a total of 4 + 7 = 11 bytes were left empty. That's why the size is not 37 but 11+37 = 48. Lmk if that helps.

antholuo commented on UART Driver/Core/Src/main.c:
```diff
@@ -0,0 +1,307 @@
+		}
+
+		//Convert the byte array into an fijo struct
+		struct fijo *ptr = (struct fijo*)buf;
```
Not sure if we want to use "ptr" or if we want a more descriptive name.

YashrajN replied:
Ya fs I changed it to a more meaningful name

antholuo reviewed (commented):
Looks good to me!
A few small comments here and there overall.

upadhyaydhruv reviewed (changes requested):
Please use gitignore or some other form of commits to only show the relevant files in the PR?

YashrajN committed:
Got rid of unnecessary files

YashrajN committed:
Update print.py

upadhyaydhruv commented on UART Driver/main.c:
```diff
@@ -0,0 +1,307 @@
+
+	//Receives a byte to check if its the same as STARTBYTE
+	uint8_t check;
+	HAL_UART_Receive(&huart2, &check, 1, HAL_MAX_DELAY);
```
Use an interrupt here. Otherwise you will have to continuously poll. If you use HAL_UART_Receive_IT, you can make it so the readFIJO() function is only called when a UART buffer is received. That would make things more efficient and less blocking. 

Also, depending on how quickly messages are sent from jetson, FW might have to implement a FIFO system where structs received from jetson are added to a queue and processed in the order they are received in. Something to consider when you integrate into telemetry manager

YashrajN replied:
Sounds good Ill keep it in mind

upadhyaydhruv commented on UART Driver/main.c:
```diff
@@ -0,0 +1,307 @@
+	uint8_t check;
+	HAL_UART_Receive(&huart2, &check, 1, HAL_MAX_DELAY);
+
+	if(check == STARTBYTE){
```
the STARTBYTE can be used as the trigger for the interrupt. Search up UART interrupts in stm32 HAL to get more insight

YashrajN replied:
Sounds good Ill do some research on it

YashrajN replied:
Ok so from what I am understanding the idea is that I don't want to keep blocking the CPU when looking for STARTBYTE or until the delay is over and that's why it's better to use the interrupt. But I am pretty sure it is only required here for STARTBYTE since I know after STARTBYTE I will be receiving my other bytes and I don't need to interrupt every time for the rest of the message but lmk what you think

upadhyaydhruv replied:
yeah so you would only call the interrupt for your start byte. Since you know the structure of your struct/data after hte start byte you can just get the data accordingly, no need to have an interrupt every time

upadhyaydhruv commented on UART Driver/main.c:
```diff
@@ -0,0 +1,307 @@
+  huart2.Init.BaudRate = 115200;
+  huart2.Init.WordLength = UART_WORDLENGTH_8B;
+  huart2.Init.StopBits = UART_STOPBITS_1;
+  huart2.Init.Parity = UART_PARITY_NONE;
```
We should add parity or a CRC check of some kind to maintain message robustness. Talk to @antholuo for more info re this, he worked on CRC

YashrajN replied:
Ok ya fs. But I was under the impression that this stuff wasn't as important to change as it's just dev code so far and independent from ZPs message robustness. I could be completely wrong on this tho so pls lmk what u think.

upadhyaydhruv replied:
Yeah definitely, this is just for general prototyping so its fine. Just something to keep in mind when integrating into Telemetry manager

upadhyaydhruv reviewed (changes requested):
Overall well done, you said you tested this already which is good to hear.

Mainly added some suggestions on how to approach the Telemetry Manager integration of this, as to make the firmware better.

Aidan-B commented on UART Driver/main.c:
```diff
@@ -0,0 +1,307 @@
+};
+
+//Should be 37 bytes but because of padding its really 48 bytes
+ struct foji{
```
Can we define foji/fijo in a comment? I have no idea what these acronyms mean.

YashrajN replied:
Ya for sure my bad.

Gfcello replied:
Flight in Jetson out? Just guessing

YashrajN replied:
Yes, I will update the PR so it's visible the second I'm finished making the other changes

Aidan-B replied:
I think your comments are backwards for the acronyms (foji != flight controller in jetson out)

YashrajN replied:
Yes, you are correct my bad I just fixed it!

Aidan-B commented on UART Driver/print.py:
```diff
@@ -0,0 +1,22 @@
+print(ser.name)
+# arr = ['$', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\xa3', 's', '~', '\x8a', '\xe3', '8', '&', '@', '.', '\xc9', '\x01', '\xbb', '\x9a', '8','6', '@', 'E', ',', 'b', '\xd8', 'a', '\xaa', '@', '@', '\xb3', '\xc3', '1', 'B', 'V', '4', '^', 'B', '}', 'R', '\x85', 'B', 'a', '\x06', '\x00', '\x08']
+# arr = [b'\x00', b'\xb8', b'\x1e', b'\x85', b'\xeb', b'Q', b'8', b'&', b'@', b'\xb8', b'\x1e', b'\x85', b'\xeb', b'Q', b'8', b'6',b'$', b'\x01', b'\x01', b'\x01', b'\x00', b'\x00', b'\x00', b'\x00', b'\xb8', b'\x1e', b'\x85', b'\xeb', b'Q', b'8', b'&', b'@', b'\xb8', b'\x1e', b'\x85', b'\xeb', b'Q', b'8', b'6', b'@']
+# ser.write(arr[0])
```
I know this is all commented out, but perhaps the fact that almost half of the code is commented out here means that it should be removed, or abstracted into a function so we only comment out a small amount. Right now, I think these massive array of chars should be explained.

YashrajN replied:
Sure sounds good, I was thinking of removing it altogether as it only makes things confusing and is not part of the driver anyway so I think I will do that, thanks.

Gfcello replied:
For the "COM5" and baude rate, declaring them as:
`SERIAL_BAUD_RATE = 115200`
and then calling 
`serial.Serial(SERIAL_PORT, SERIAL_BAUD_RATE)`
makes the code much cleaner and easier to understand

Gfcello replied:
To me the "arr" variable looks like either testing code or an expected output, either way it can be removed

YashrajN replied:
Ya I agree this file is definitely messy and not required it was only meant for testing purposes.

Aidan-B reviewed (commented):
I think it might also be valuable if you checkout confluence and update/create any documentation to support this PR. There is a page on UART, and if you know anything that isn't on that page, I think it would be valuable if you update it.

Gfcello commented on UART Driver/main.c:
```diff
@@ -0,0 +1,307 @@
+const uint8_t STARTBYTE = 0x24;
+
+void sendFOJI(){
+	//Get out from telemetry manager
+	struct foji out;
+
+	HAL_UART_Transmit(&huart2, &out, sizeof(struct foji), HAL_MAX_DELAY);
```
Maybe name the FOJI struct to something more meaningful, like pogi_msg (Plane out ground in).

I like the use of sizeof especially for this when the struct size might change!

YashrajN replied:
LOL thanks I was initially using my estimated struct size and it screwed everything over cuz I wasn't taking into account the padding, and ya fs I was thinking of making a comment above the struct explaining what it means.

Gfcello commented on UART Driver/main.c:
```diff
@@ -0,0 +1,307 @@
+	HAL_UART_Receive(&huart2, &check, 1, HAL_MAX_DELAY);
+
+	if(check == STARTBYTE){
+		//Receive all the message bytes in buf
+		for(uint16_t i = 1; i < sizeof(struct fijo); i++)
+		{
+			HAL_UART_Receive(&huart2, &buf[i], 1, HAL_MAX_DELAY);
+		}
```
I believe that there is a way to read in a number of bytes without needing a loop. Think you pass in the sizeof(struct fijo) instead of a 11 in the UART_Receive function, but that would be more ideal than using a loop. Be careful about whether the data coming in is stored in little endian or big endian format (order of bytes).

YashrajN replied:
I was thinking about this too but in this case, I would need to start receiving from index 1 and not 0. So if I put the address of my array and the size of the struct as my parameter it will mess it up cuz it starts from the 0 index

Gfcello replied:
Could you add a single byte to your struct of padding? It seems like that byte gets transmitted anyway.

YashrajN replied:
Hmm, I am not sure I understand what you mean. But do you mean like adding it when I cast my byte array to a struct?

Gfcello replied:
Adding it at the start to solve that indexing issue. Then the data you care about can be read from the struct and that one byte will be ignored.

YashrajN replied:
oh like getting rid of the "start" uint8_t from the fijo struct altogether because it's not functional? that would work

Gfcello commented on UART Driver/main.c:
```diff
@@ -0,0 +1,307 @@
+
+  /* Infinite loop */
+  /* USER CODE BEGIN WHILE */
+  while (1)
+  {
+    /* USER CODE END WHILE */
+
+    /* USER CODE BEGIN 3 */
+	  sendFOJI();
+	  readFIJO();
+  }
```
As stated earlier, having this be an interrupt prevents this from blocking.

Is there a frequency of data transmission for this and is it handled elsewhere? This currently will send messages as fast as it can. You could use a timer to set the data frequency or use the telemetry manager's output rate to set it (Not sure how accurate that is). 
If the send function is called after each read function interrupt triggers than the output would be reasonably consistent.

YashrajN replied:
I think I would call the send function when required I just put it in the loop for sake of putting it in the loop. This would most likely be adjusted when it is implemented in telemetry manager.

Gfcello replied:
Makes sense, this way it gets called lots for debugging

Gfcello commented on UART Driver/main.c:
```diff
@@ -0,0 +1,307 @@
+	//Initializes starting byte
+	buf[0] = STARTBYTE;
+
+	struct fijo in;
+	in.start = STARTBYTE;
```
Making this name a little more descriptive eg. jetson_output, or msg_from_jetson. something like that would make this clearer.

Gfcello commented on UART Driver/main.c:
```diff
@@ -0,0 +1,307 @@
+	struct gpsCoordinatesFIJO gpsCoord;
+};
+
+const uint8_t STARTBYTE = 0x24;
+
```
Veery nitpicky, but constants are easier read with a _ in the name, so START_BYTE

YashrajN committed:
Review Changes

-Updated the constant and variable names

-Got rid of the unnecessary files

-Implemented the UART Interrupt but it makes it so the byte reading is inconsistent

Next steps:
-Simplifying the fijo reception for loop to a single line (will be done once testing is finished)

-Create documentation for the driver (but will do that once merged)

YashrajN committed:
Update main.c

YashrajN committed:
Update main.c

YashrajN committed:
Update main.c

YashrajN committed:
Update main.c

YashrajN committed:
Implemented UART Interrupt

UART interrupt is now functional
Unnecessary loops removed

YashrajN committed:
Updated the encoding protocol

YashrajN committed:
Struct Simplification

Got rid of the extra structs since they had no purpose and were just confusing
